### PR TITLE
Added error wrapping options to `race`

### PR DESCRIPTION
**Pull request by Lucretiel**
Added features described in #973

@Andarist

### CODE REVIEW

File: src/internal/proc.js [hunk 1]
```diff
@@ -600,9 +600,15 @@ export default function proc(
         }
 
         if(isErr) {
+          // TODO: pre-calculate these switches
+          // TODO: allow users to customize origin / raceOrigin key?
+          // TODO: is this behavior correct if typeof res !== Error?
+          if(annotateErr) { res.raceOrigin = key }
```
Comment (Andarist) [inline src/internal/proc.js]: Thinking about the name for this - any other ideas than `raceOrigin`, `effectOrigin` ?

Comment (Andarist) [inline src/internal/proc.js]: also seting this property should be done with `defineProperty` I think and it should be excluded from enumerability

**Comment by Andarist** [comment]
I was thinking more about adding the second option (`annotateErr`) as default and for everyone. Dont think there is actually need for `options` (at least not yet).

This would allow keeping the code simpler.

**Comment by Lucretiel** [comment]
I wasn't sure if modifying the `Error` counted as a breaking change; I also wasn't sure about the case where an Effect throws something that's not an `Error`

File: src/internal/proc.js (hunk 1 above)
Comment (Lucretiel) [inline src/internal/proc.js]: I'll update the diff.

> any other ideas than `raceOrigin`, `effectOrigin`

Nothing comes to mind... any reason?

**Comment by Andarist** [comment]
good point about throwing something else than `Error`, we could check `instanceof` before assigning this prop

as to modifying the `Error` - should be no biggy I think

File: src/internal/proc.js (hunk 1 above)
Comment (Andarist) [inline src/internal/proc.js]: not yet, will try to think about the most appropriate name

**Comment by Lucretiel** [comment]
How will users determine the origin of non-`Error` errors? For instance, jqXHR just raises itself in the event of an error. Can we just expect users just wrap these in `Error` types?

**Comment by Andarist** [comment]
`jqXHR` does not implement Promise/A+ specification, thus we won't support its non-standard behaviours.

If you want to benefit from this change while using `jqXHR` I would advise you to write a wrapper for making requests which should use this to pass the `Error` to the saga:
```javascript
function request(opts) {
	var promise = $.ajax(opts)
		.then(
		data => data,
		(jqXHR, textStatus, errorThrown) => {
			// use those ^ to construct the Error
			return new Error()
		})
	
	return promise
}
```
with that approach `redux-saga` will be able to handle both cases by chaining `.then` to the returned `promise`

File: src/internal/proc.js (hunk 1 above)
Comment (Lucretiel) [inline src/internal/proc.js]: Do you mean that you want it to be more unique? Like `CANCEL`?

**Comment by Lucretiel** [comment]
How would you feel about retaining `wrapErr`, in that case? That way, users have a catch-all solution if they want to catch thrown objects more generically and immutably.

**Comment by Andarist** [comment]
Its not really immutable way of wrapping the error - it mutates the original one, but I dont see anything wrong with that in this particular use case.

Would like to keep this change minimal and adding `options` to the race is quite a big change, no other effect is configurable like that.

**Comment by Lucretiel** [comment]
It seems to me that there are effects taking more than one argument, but I take your meaning about wanting to keep things simple. Will fix.

File: src/internal/proc.js (hunk 1 above)
Comment (Lucretiel) [inline src/internal/proc.js]: How would you feel about a new symbol, then? Users would do `err[RACE_ORIGIN]` instead of `err.raceOrigin`

**Comment by Lucretiel** [comment]
Also, I'm trying to write a test for this functionality, but I confess I cannot for the life of me understand what the existing tests are actually doing.

**Comment by Andarist** [comment]
> It seems to me that there are effects taking more than one argument, but I take your meaning about wanting to keep things simple. Will fix.

Sure there are, I ment that there is no 'configurable' effect in such a way now. My comment was not about arguments count per se, just presence of `options` argument

> Also, I'm trying to write a test for this functionality, but I confess I cannot for the life of me understand what the existing tests are actually doing.

Could u give a link to the lines you have trouble with? I'm happy to explain everything

**Comment by Lucretiel** [comment]
I added a test, so you can tell me if it makes sense or not. Mostly I'm not clear about what the Promise.resolve is doing at the beginning of the test.

File: docs/api/README.md [hunk 2]
```diff
@@ -37,7 +37,7 @@
   * [`setContext(props)`](#setcontextprops)
   * [`getContext(prop)`](#getcontextprop)
 * [`Effect combinators`](#effect-combinators)
-  * [`race(effects)`](#raceeffects)
+  * [`race(effects, options)`](#raceeffects)
```
Comment (Andarist) [inline docs/api/README.md]: this should be removed

File: docs/api/README.md [hunk 3]
```diff
@@ -761,12 +761,14 @@ Creates an effect that instructs the middleware to return a specific property of
 
 ## Effect combinators
 
-### `race(effects)`
+### `race(effects, options)`
```
Comment (Andarist) [inline docs/api/README.md]: and this too

File: docs/api/README.md [hunk 4]
```diff
@@ -761,12 +761,14 @@ Creates an effect that instructs the middleware to return a specific property of
...
 
 Creates an Effect description that instructs the middleware to run a *Race* between
-multiple Effects (this is similar to how [`Promise.race([...])`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise/race) behaves).
+multiple Effects (this is similar to how [`Promise.race([...])`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise/race) behaves). When one effect completes or throws, that result is returned (or thrown) by the race, and all the other effects are cancelled. If the result is not an error, it wrapped in an object of the form `{[key]: result}`; this allows the caller to determine which Effect returned the result. If the effect throws an `Error`, the `raceOrigin` property is attached to the thrown error, allowing the caller to determine which Effect threw the error.
```
Comment (Andarist) [inline docs/api/README.md]: this should be revamped and replaced with `RACE_ORIGIN` mention and example

File: src/internal/proc.js [hunk 5]
```diff
@@ -1,4 +1,8 @@
-import { noop, kTrue, is, log as _log, check, deferred, uid as nextEffectId, array, remove, object, TASK, CANCEL, SELF_CANCELLATION, makeIterator, createSetContextWarning, deprecate, updateIncentive } from './utils'
+import {
```
Comment (Andarist) [inline src/internal/proc.js]: hm, thats becoming quite a long list, but newline shouldnt be inserted here yet, gonna think how those utils could be splitted later

File: src/internal/proc.js [hunk 6]
```diff
@@ -600,6 +604,13 @@ export default function proc(
         }
 
         if(isErr) {
+          // TODO: allow users to customize origin / raceOrigin key?
```
Comment (Andarist) [inline src/internal/proc.js]: those comments are really not neccessary

File: test/proc/race.js [hunk 7]
```diff
@@ -110,3 +110,22 @@ test('processor race between sync effects', assert => {
     assert.end();
   })
 });
+
+test('processor race error has origin', assert => {
```
Comment (Andarist) [inline test/proc/race.js]: most tests do not put assert calls inside a generator, but rather have `actual` and `expected` arrays which are compared at the end, so basically that would look something like
```
test('processor race error has origin', assert => {
  assert.plan(1);
  const actual = []
  const expected = ['willErr']

  function* genFn() {
    try {
      yield io.race({
        wontResolve: new Promise((resolve, reject) => {}),
        willErr: Promise.reject(new Error("FAIL")),
      })
    } catch(err) {
      actual.push(err[RACE_ORIGIN])
    }
  }

  proc(genFn()).done.then(() => {
    assert.deepEqual(actual, expected, 'thrown error should have RACE_ORIGIN set on itself')
  })
})
```

sometimes smth like `Promise.resolve(1)` is used in test to cause asynchrony, no need for that here though

this should also be accompanied by a test which do not set `RACE_ORIGIN`, when something else than an Error is thrown

**Comment by Andarist** [comment]
Sorry for a long delay with this PR, but I've decided not to proceed with this feature right now. If you want to know error's origin you can track it on your own with a simple helper.

**Pull request closed without merging**